Turbulent channel flow

jbrillon opened the pull request:
PR to incorporate the `channel_flow` flow case and the features associated with it. 

_Description to be updated, waiting for PR #171 and PR #172 to further develop this branch _

jbrillon commented on src/dg/dg.cpp:
```diff
@@ -434,6 +434,54 @@ void DGBaseState<dim,nstate,real,MeshType>::update_model_variables()
     pde_model_rad_fad->cellwise_volume.update_ghost_values();
 }
 
+template <int dim, int nstate, typename real, typename MeshType>
```
Since these values are computed in `FlowSolver` through a `FlowSolverCaseBase` object and need to be assigned to the `ModelBase` object held in DG, I chose to implement a setter function in DG that we can call from within `FlowSolver` which has ; this is a different scenario than calling the "update_model_variables()" function since that function cannot compute quantities that are related to a specific physics or flow solver case

sivanadarajah commented on src/physics/navier_stokes.cpp:
```diff
@@ -116,6 +116,47 @@ ::compute_temperature_gradient (
+        velocity_gradient_of_parallel_velocity_in_the_direction_normal_to_wall += velocities_gradient[0][d]*normal_vector[d];
+    }
+    // Reference: https://www.cfd-online.com/Wiki/Wall_shear_stress
+    const real2 wall_shear_stress = scaled_viscosity_coefficient*velocity_gradient_of_parallel_velocity_in_the_direction_normal_to_wall;
```
Is this computed in the computational or physical domain?

jbrillon replied:
physical

jbrillon commented:
Resolved all outdated comments; will be doing a self-review before marking this ready for review

jbrillon commented on src/dg/dg_base.cpp:
```diff
@@ -1282,12 +1277,11 @@ void DGBase<dim,real,MeshType>::assemble_residual (const bool compute_dRdW, cons
-    } catch(...) {
-        assembly_error = 1;
-    }
+    // } catch(...) {
+    //     assembly_error = 1;
+    // }
```
without this turned off, its impossible to debug the code -- let me know how you think we should handle this

cpethrick replied:
From discussion a few weeks ago, there seemed to be a way to get around this. Could you test that and (if successful), revert this?

jbrillon replied:
@dougshidong what was the command you mentioned?

jbrillon replied:
Reverted the changes and added a comment

jbrillon commented on src/dg/dg_factory.cpp:
```diff
@@ -67,12 +70,36 @@ ::create_discontinuous_galerkin(
+                using SGS_enum = Parameters::PhysicsModelParam::SubGridScaleModel;
+                const SGS_enum SGS_model_type = parameters_input->physics_model_param.SGS_model_type;
+                if(SGS_model_type == SGS_enum::shear_improved_smagorinsky) {
+                    // this->do_compute_mean_strain_rate_tensor = true;
+                    // return std::make_shared< DGStrongLES_ShearImproved<dim,dim+2,real,MeshType> >(parameters_input, degree, max_degree_input, grid_degree_input, triangulation_input);
+                    // TO DO: Create DGStrongLES_Base
+                    // TO DO: Create DGStrongLES_ConstantModelCoefficient
+                    // TO DO: Create DGStrongLES_ShearImproved
```
address this

jbrillon commented on src/parameters/parameters_physics_model.cpp:
```diff
@@ -45,6 +53,28 @@ void PhysicsModelParam::declare_parameters (dealii::ParameterHandler &prm)
+                              dealii::Patterns::Double(1e-15, dealii::Patterns::Double::max_double_value),
+                              "Clipping limit for the Dynamic Smagorinsky model constant (default is 0.01).");
+
+            prm.declare_entry("apply_low_reynolds_number_eddy_viscosity_correction", "false",
+                              dealii::Patterns::Bool(),
+                              "Flag for applying the low Reynolds number eddy viscosity correction. By default, false.");
+
```
add something in modify parameters for setting this to true for a given Reynolds number

jbrillon commented on src/physics/model_factory.cpp:
```diff
@@ -122,6 +243,37 @@ ::create_Model(const Parameters::AllParameters *const parameters_input)
                 return nullptr;
             }
+        }
+        // -------------------------------------------------------------------------------
+        // Navier-Stokes model
+        // -------------------------------------------------------------------------------
+        else if (model_type == Model_enum::navier_stokes_model) {
+            if constexpr ((nstate==dim+2) && (dim==3)) {
+                // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+                // Navier-Stokes with model source terms (e.g. channel flow)
+                // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+                return std::make_shared < NavierStokesWithModelSourceTerms<dim,nstate,real> > (
+                    parameters_input,
+                    parameters_input->euler_param.ref_length,
+                    parameters_input->euler_param.gamma_gas,
+                    parameters_input->euler_param.mach_inf,
+                    parameters_input->euler_param.angle_of_attack,
+                    parameters_input->euler_param.side_slip_angle,
+                    parameters_input->navier_stokes_param.prandtl_number,
+                    parameters_input->navier_stokes_param.reynolds_number_inf,
+                    parameters_input->navier_stokes_param.use_constant_viscosity,
+                    parameters_input->navier_stokes_param.nondimensionalized_constant_viscosity,
+                    parameters_input->navier_stokes_param.temperature_inf,
+                    parameters_input->flow_solver_param.relaxation_coefficient_for_turbulent_channel_flow_source_term,
+                    parameters_input->navier_stokes_param.nondimensionalized_isothermal_wall_temperature,
+                    parameters_input->navier_stokes_param.thermal_boundary_condition_type,
+                    manufactured_solution_function,
+                    parameters_input->two_point_num_flux_type);
```
is this still used?

jbrillon commented on src/physics/physics_factory.cpp:
```diff
@@ -138,7 +138,56 @@ ::create_Physics(const Parameters::AllParameters                              *c
+        }
+    } else if (pde_type == PDE_enum::navier_stokes_channel_flow_constant_source_term_wall_model) {
+        if constexpr (nstate==dim+2) {
+            const double domain_length_y_direction = parameters_input->flow_solver_param.turbulent_channel_domain_length_y_direction;
+            const double half_channel_height = domain_length_y_direction/2.0;
+            const int number_of_cells_y_direction = parameters_input->flow_solver_param.turbulent_channel_number_of_cells_y_direction;
+            const double uniform_spacing_y_direction = domain_length_y_direction/double(number_of_cells_y_direction);
```
note: hard coded for turbulent channel flow case / uniform grids

jbrillon replied:
add abort if flow case type is not channel flow

jbrillon commented on tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_tgv_strong_dg_adaptive_time_step_from_restart.sh:
remove this?

jbrillon commented on tests/integration_tests_control_files/taylor_green_vortex_integration/viscous_tgv_strong_dg_adaptive_time_step_from_restart.sh:
for the reviewers: I simply changed the permissions of this file; that's why it shows as empty

jbrillon commented on tests/unit_tests/tke_spectra_calculation_fix/tke_spectra_calculation_fix.cpp:
```diff
@@ -0,0 +1,66 @@
+    std::unique_ptr<PHiLiP::FlowSolver::PeriodicTurbulence<dim, nstate>> periodic_turbulence = std::make_unique<PHiLiP::FlowSolver::PeriodicTurbulence<dim,nstate>>(&all_parameters);
+    // periodic_turbulence->output_velocity_field(flow_solver->dg,0,8.0);
+
+    // do it again for the next outputted velocity field
+    periodic_turbulence->output_velocity_field(flow_solver->dg,1,9.0);
+    // pcout << "reading values from file prefix: \n " << input_filename_prefix << " \n and projecting... " << std::flush;
```
improve this so that the `1` is not hard-coded

jbrillon commented on tests/integration_tests_control_files/dipole_wall_collision/input.prm:
add a test associated with this file

jbrillon commented on src/physics/rough_code_notes_for_implementing.cpp:
remove this file

jbrillon replied:
I have sent it to Dominic

jbrillon commented on src/physics/physics_model.h:
```diff
@@ -135,6 +194,56 @@ class PhysicsModel : public PhysicsBase <dim, nstate, real>
+     *  Solution: In order to make the hidden function visible in derived class, 
+     *  we need to add the following:
+    */
+    // using PhysicsBase<dim,nstate,real>::dissipative_flux; // can delete if the arguments of this function are updated to match PhysicsBase with filtered solution being passed
+    using PhysicsBase<dim,nstate,real>::boundary_face_values; // can delete if the arguments of this function are updated to match PhysicsBase with filtered solution being passed
+    using PhysicsModel<dim,nstate,real,nstate_baseline_physics>::dissipative_flux;
```
look into this

jbrillon commented on src/physics/physics_model.h:
```diff
@@ -13,6 +13,16 @@ namespace Physics {
+     *  Solution: In order to make the hidden function visible in derived class, 
+     *  we need to add the following:
+    */
+    using PhysicsBase<dim,nstate,real>::dissipative_flux; // can delete if the arguments of this function are updated to match PhysicsBase with filtered solution being passed
+    using PhysicsBase<dim,nstate,real>::boundary_face_values; // can delete if the arguments of this function are updated to match PhysicsBase with filtered solution being passed
```
look into this

jbrillon commented on src/physics/navier_stokes.cpp:
```diff
@@ -956,13 +1406,390 @@ ::boundary_manufactured_solution (
+            conservative_soln[s] = uh(s);
+        }
+        const std::array<double, nstate> primitive_soln = this->template convert_conservative_to_primitive_templated<real>(conservative_soln);
+        // if (primitive_soln[0] < 0) this->pcout << evaluation_points << std::endl;
```
delete this line

jbrillon commented on src/physics/large_eddy_simulation.h:
```diff
@@ -149,6 +176,13 @@ class LargeEddySimulationBase : public ModelBase <dim, nstate, real>
         const dealii::types::global_dof_index cell_index) const;
+
+    /// Channel flow source term
+    /** Forcing function to maintain the expected bulk Reynolds number throughout the solution
```
confirm this

jbrillon commented on src/physics/navier_stokes_model.h:
```diff
@@ -0,0 +1,111 @@
+namespace PHiLiP {
+namespace Physics {
+
+/// Large Eddy Simulation equations. Derived from Navier-Stokes for modifying the stress tensor and heat flux, which is derived from PhysicsBase. 
```
update comment

jbrillon commented on src/physics/navier_stokes_model.cpp:
```diff
@@ -0,0 +1,190 @@
+        std::abort();
+    }
+    // x-momentum term
+    const real bulk_reynolds_number = this->navier_stokes_physics->reynolds_number_inf;
+    const real viscosity_coefficient = this->navier_stokes_physics->constant_viscosity;
+    const real scaled_viscosity_coefficient = this->navier_stokes_physics->scale_viscosity_coefficient(viscosity_coefficient);
+    const real expected_mass_flow_rate = scaled_viscosity_coefficient * bulk_reynolds_number / this->half_channel_height;
+    source_term[1] = this->resultant_wall_shear_force/this->domain_volume - this->relaxation_coefficient*(this->bulk_mass_flow_rate - expected_mass_flow_rate)/this->time_step;
+
```
review this

jbrillon commented on src/parameters/parameters_flow_solver.cpp:
```diff
@@ -130,6 +136,11 @@ void FlowSolverParam::declare_parameters(dealii::ParameterHandler &prm)
                           dealii::Patterns::Double(0,dealii::Patterns::Double::max_double_value),
                           "Outputs the restart files at time intervals of dt.");
+        
+        prm.declare_entry("write_unsteady_data_table_file_every_dt_time_intervals", "0.0",
```
should set the default value to `0.1` and set it to zero explicitly in certain test cases with restart files

jbrillon commented on src/parameters/parameters_flow_solver.cpp:
```diff
@@ -258,7 +269,79 @@ void FlowSolverParam::declare_parameters(dealii::ParameterHandler &prm)
                               "Flag to calculate numerical entropy and write to file. By default, do not calculate.");
-            
+
+            prm.declare_entry("check_nonphysical_flow_case_behavior", "false",
```
should set to true if it doesnt break any tests

jbrillon commented on src/parameters/parameters_flow_solver.cpp:
```diff
@@ -258,7 +269,79 @@ void FlowSolverParam::declare_parameters(dealii::ParameterHandler &prm)
+                              " <laminar | "
+                              " manufactured | "
+                              " turbulent>.");
+            prm.declare_entry("relaxation_coefficient_for_turbulent_channel_flow_source_term", "0.0",
```
remove this?

jbrillon commented on src/parameters/all_parameters.cpp:
```diff
@@ -274,15 +285,20 @@ void AllParameters::declare_parameters (dealii::ParameterHandler &prm)
                       "  euler | "
                       "  mhd |"
                       "  navier_stokes |"
+                      "  navier_stokes_channel_flow_constant_source_term | "
```
should add the option if doing constant or unsteady source term for the channel flow with default as the constant one

jbrillon commented on src/parameters/all_parameters.cpp:
```diff
@@ -578,59 +668,41 @@ const std::string test_string = prm.get("test_type");
+      
+      pcout << "done." << std::endl;
+    } 
+    else if((pde_type == PartialDifferentialEquation::navier_stokes || 
+             (pde_type == PartialDifferentialEquation::physics_model || pde_type == PartialDifferentialEquation::physics_model_filtered)) && 
+              (flow_solver_param.flow_case_type == flow_solver_param.FlowCaseType::taylor_green_vortex || 
+                flow_solver_param.flow_case_type == flow_solver_param.FlowCaseType::decaying_homogeneous_isotropic_turbulence)) {
```
need to add the channel flow stuff in here

jbrillon commented on src/dg/strong_dg_les.cpp:
```diff
@@ -0,0 +1,1076 @@
+    // allocate/reinit the model variables
+    allocate_model_variables();
+
+    // TO DO: should only call this once if no hp-adaptation
+    update_cellwise_volume_and_poly_degree();
```
for the reviewers: this could be addressed in a future PR -- I only left this comment because it would add some efficiency to the code, but is not currently essential

jbrillon replied:
left a comment in the code

jbrillon reviewed (commented):
For the reviewers: I have completed a self-review of all the changes in this PR and have cleaned it up ensuring that it is up to standards. Given the size of this PR, I recommend only carefully reviewing changes/files that are relevant to your own work / that might affect your cases. As you will see, most of this PR is additions, and most of these are contained to themselves by flags or separate classes. While you may see some "TO DO:" comments in `channel_flow.cpp`, these will eventually get addressed or removed in a future PR. I've decided to open this PR for review since I believe more than a few students will benefit from having these improvements/features on the master branch.

jbrillon commented on src/physics/physics_model.cpp:
```diff
@@ -291,6 +412,55 @@ ::max_viscous_eigenvalue (const std::array<real,nstate> &/*conservative_soln*/)
+        physics_baseline->boundary_face_values(
+                boundary_type, pos, normal_int, baseline_soln_int, baseline_soln_grad_int, 
+                baseline_soln_bc, baseline_soln_grad_bc);
+        /*TO DO NOT IMPLEMENTED YET
+        model->boundary_face_values(
+                boundary_type, pos, normal_int, soln_int, soln_grad_int, 
+                soln_bc, soln_grad_bc);*/
+
+        for(int s=0; s<nstate_baseline_physics; ++s){
+            soln_bc[s] += baseline_soln_bc[s];
+            soln_grad_bc[s] += baseline_soln_grad_bc[s];
+        }
+        std::abort(); // <-- TO DO NOT IMPLEMENTED YET
```
For the reviewers: the only case this would hit is for RANS, which I am not sure what it should be. Therefore I have added an abort if trying to run RANS for a flow with a wall boundary condition.

jbrillon replied:
Added a print statement before all aborts in this file

cpethrick commented:
It looks like there's a lot of useful additions in this PR! Looking forward to having this in master. Great work!

I have a few general questions/suggestions before I go into a detailed review:
* Can you confirm whether `ctest` is passing as expected? How long are the new tests?
* Can you add labels to your new tests per @tofstie 's PR #269? 
* (Question for all reviewers any maybe better discussed in-person): how do we want to split up the PR review so that we ensure all files are looked at while getting reviews through in a reasonable amount of time?

PranshulThakur commented:
Added @DominicRoy2 to review SGS and wall models as he would be using the code in this PR extensively in the future.

jbrillon commented:
> It looks like there's a lot of useful additions in this PR! Looking forward to having this in master. Great work!
> 
> I have a few general questions/suggestions before I go into a detailed review:
> 
>     * Can you confirm whether `ctest` is passing as expected? How long are the new tests?
> 
>     * Can you add labels to your new tests per @tofstie 's PR [Adding Labels to Ctests #269](https://github.com/dougshidong/PHiLiP/pull/269)?
> 
>     * (Question for all reviewers any maybe better discussed in-person): how do we want to split up the PR review so that we ensure all files are looked at while getting reviews through in a reasonable amount of time?

Thank you! 
All the tests I've added are super quick (about 1 minute), I've added all the ctest labels to my tests in the last merge commit, and I haven't ran a full ctest yet, was planning to do that when it's closer to being merged since it takes around 24hours to run, but every test (new and old) related to TGV passes.

jbrillon commented:
> Added @DominicRoy2 to review SGS and wall models as Dominic would be using the code in this PR extensively in the future.

Ok sounds good!

cpethrick commented on src/physics/navier_stokes.cpp:
```diff
@@ -55,67 +57,177 @@ NavierStokes<dim, nstate, real>::NavierStokes(
+template <int dim, int nstate, typename real>
+template<typename real2>
+real2 NavierStokes<dim,nstate,real>
+::compute_wall_shear_stress (
```
Sounds good, but could you add an additional comment explaining that in the code?

jbrillon replied:
I've added this comment inside the function before the big block of commented code

cpethrick commented on src/physics/navier_stokes.cpp:
```diff
@@ -365,8 +625,9 @@ ::compute_deviatoric_strain_rate_tensor (
-            deviatoric_strain_rate_tensor[d1][d2] = strain_rate_tensor[d1][d2] - (1.0/3.0)*vel_divergence;
+            deviatoric_strain_rate_tensor[d1][d2] = strain_rate_tensor[d1][d2];
         }
+        deviatoric_strain_rate_tensor[d1][d1] -= (1.0/3.0)*vel_divergence;
```
This looks different from before. Are there any impacts on results using NS without this change?

jbrillon replied:
It only corrects a specific post processed quantity in the TGV case, so no it wont impact anyone's results -- it corrects the viscous dissipation rate in the KE budget which no one else in the lab is looking at

cpethrick replied:
Sounds good. Thanks for clarifying!

cpethrick commented on src/physics/navier_stokes.cpp:
```diff
@@ -956,13 +1417,406 @@ ::boundary_manufactured_solution (
+    }
+    names.push_back ("total_energy");
+    names.push_back ("pressure");
+    names.push_back ("pressure_coeffcient");
```
this could be an opportunity to correct this typo (`pressure_coefficient`). But I'm not sure if that would mess up anyone's paraview scripts...

jbrillon replied:
Sure I can fix it; doesn't break the code so why not

jbrillon replied:
fixed

cpethrick commented on src/flow_solver/flow_solver_cases/channel_flow.cpp:
I saw your comment that the `TO-DO` items will get addressed in the future. That's fine by me, but could you open an issue about that once this PR is close to being merged?

jbrillon replied:
They aren't needed, they're just notes for myself so I'll just remove them

jbrillon replied:
fixed

cpethrick commented on src/flow_solver/flow_solver_cases/dipole_wall_collision.cpp:
```diff
@@ -0,0 +1,225 @@
+    const int number_of_cells_x_direction = this->number_of_cells_per_direction;
+    const int number_of_cells_y_direction = this->number_of_cells_per_direction;
+    // ------------------------------------------------
+    // CODE BELOW TAKEN FROM CHANNEL FLOW CLASS -- Commented z-component stuff since 2D here
```
Could you move the grid generation here and in channel flow to the `src/mesh/grids` directory and merge this with `turbulent_channel_flow`? Might be some additional work to generalize dimensions, stretch factor, which face is refined etc but it could be useful for the future.

jbrillon replied:
I thought of that, however these mesh stretching functions are specific to the turbulent channel flow / dipole case so I chose to leave them in here; I think could be done by in a future PR by someone else that wants to benefit from this meshing code and use it in their case

cpethrick commented on src/flow_solver/flow_solver_cases/dipole_wall_collision.h:
```diff
@@ -0,0 +1,59 @@
+                        const bool is_oblique_=false);
+
+    /// Destructor
+    ~DipoleWallCollision() {};
```
I would advocate for removing default destructors throughout the PR for consistency with the rest of the code.

PranshulThakur replied:
Agreed, except the base class destructors that need to be declared virtual. I had run into some memory issues due to that in the past.

jbrillon replied:
Will be addressed in issue #317 , I've marked it as optional since it's not critical for this PR to go through

cpethrick reviewed (commented):
I have looked over a few files and have some minor comments. I will plan to review the code in chunks and release comments as I go so that others can see my comments sooner.

Everything is very clean, thanks for that!

jbrillon commented on src/physics/mhd.cpp:
```diff
@@ -74,6 +74,24 @@ ::convert_primitive_to_conservative ( const std::array<real,nstate> &primitive_s
+    const std::array<real,nstate> &/*primitive_soln*/,
+    const std::array<dealii::Tensor<1,dim,real>,nstate> &primitive_soln_gradient) const
+{
+    return primitive_soln_gradient;
```
yeah I can do that; these were added so that you can modally filter the gradients in strong DG but that is really only done for EULER/NS/LES

jbrillon replied:
fixed

jbrillon commented on src/physics/initial_conditions/initial_condition_function.cpp:
```diff
@@ -17,17 +17,253 @@ ::InitialConditionFunction ()
+
+template <int dim, int nstate, typename real>
+real InitialConditionFunction_NavierStokesBase<dim,nstate,real>
+::convert_primitive_to_conversative_value(
```
`this->euler_physics->convert_primitive_to_conversative_value()` doesn't exist; the function I added calls the existing function `this->euler_physics->convert_primitive_to_conservative`

cpethrick replied:
I understand the disctinction now. The naming was what confused me. Maybe something like `conservative_state_value_at_point` (or similar) would be clearer.

jbrillon replied:
It follows from the `value` virtual function that the base class `InitialConditionFunction` expects each derived class to provide. The functions `value` and `convert_primitive_to_conversative_value` are both computed point-wise / have the same arguments (`point` and `istate`) so I think its fine to leave the consistent naming as is

cpethrick commented on tests/integration_tests_control_files/dipole_wall_collision/input.prm:
Could you add `CMakeLists` files for this and the DHIT? You could use the `EXTRA-LONG` label from the recent PR if runtime is your reason for excluding these from ctest.

I'd also recommend changing this file name to something descriptive!

jbrillon replied:
Once I've submitted my thesis, I'll add a test that checks the palinstrophy value after a few time steps; similar to the TGV integration test. I'll do the same for the channel flow

jbrillon replied:
Will be resolved in issue #317

cpethrick commented on tests/integration_tests_control_files/TGV_scaling/3D_TGV_scaling_cons.prm:
```diff
@@ -37,7 +37,7 @@ set use_weight_adjusted_mass = true
-set use_periodic_bc = true
+set all_boundaries_are_periodic = true
 
 set use_classical_FR = false
```
The `use_classical_FR` parameter should be removed here and in other tests.

PranshulThakur commented on src/flow_solver/flow_solver.cpp:
```diff
@@ -588,9 +631,26 @@ int FlowSolver<dim,nstate>::run() const
+        const double avg_cpu_time_per_time_step = cpu_time/number_of_time_steps;
+        const double avg_total_wall_time_per_time_step = total_wall_time/number_of_time_steps;
+        pcout << "Elapsed CPU time: " << cpu_time << " seconds." << std::endl;
+        pcout << "Elapsed total wall time (mpi max): " << total_wall_time << " seconds." << std::endl;
```
Should the pcout comment be mpi sum?

PranshulThakur commented on src/physics/reynolds_averaged_navier_stokes.h:
```diff
@@ -56,6 +56,17 @@ class ReynoldsAveragedNavierStokesBase : public ModelBase <dim, nstate, real>
+        const bool on_boundary,
+        const dealii::types::global_dof_index cell_index,
+        const dealii::Tensor<1,dim,real> &normal,
+        const int boundary_type) const;
```
```suggestion
        const int boundary_type) const override;
```

PranshulThakur reviewed (commented):
It's a very useful PR with several significant features that would be used by a lot of people at some point in their research. Thanks for creating it and for the detailed description. Despite being long, the PR was easy to go through. 

I have added some comments, most of which are editorial while a few are on viscous BCs.

cpethrick reviewed (commented):
Added a few more comments. This time my comments are focused on the LES portions of the code.

jbrillon commented on src/dg/strong_dg.cpp:
```diff
@@ -1322,7 +1477,28 @@ void DGStrong<dim,nstate,real,MeshType>::assemble_boundary_term_strong(
 {
-    (void) current_cell_index;
+    // Get opposite face index
+    const int opposite_iface = (iface == 0) ? 1 : (
```
I think we should leave as is, since this hasn't been tested for gmesh meshes and is likely something that will get tested later on

DominicRoy2 replied:
Thanks for pointing this out. This should not pose any issues with gmsh meshes, as the face numbers do not change (the orientation of the normal does). In other words, it should still point to correct (opposite) face. I will just need to make sure they are projected correctly when I add my PR.

jbrillon commented on src/physics/navier_stokes.h:
```diff
@@ -349,10 +491,172 @@ class NavierStokes : public Euler <dim, nstate, real>
+    /// Distance from wall for wall model input velocity
+    const double distance_from_wall_for_wall_model_input_velocity;
+
+    std::unique_ptr < WallModelLookUpTable<real> > wall_model_look_up_table;
```
done

jbrillon commented on src/dg/strong_dg_les.cpp:
```diff
@@ -0,0 +1,1076 @@
+
+            std::fill(soln_at_q.begin(), soln_at_q.end(), 0.0);
+            // for (int s=0; s<nstate; ++s) {
+            //     for (int d=0; d<dim; ++d) {
```
I prefer to leave it in case it is needed in the future for future development

jbrillon commented on src/dg/strong_dg.cpp:
```diff
@@ -942,6 +950,148 @@ void DGStrong<dim,nstate,real,MeshType>::assemble_volume_term_strong(
         }
     }
 
+    // -- Solution at legendre poly
```
There are some documentation comments in lines 31-33 of strong_dg.hpp; in the current version of the code it pertains only to certain SGS models but could very well be used for anything else; the default value is false, I think this is explained in the AllParameters class

jbrillon replied:
I've added comment `// NOTE: This only pertains to advanced SGS models for LES` within the if statement

jbrillon commented on src/physics/initial_conditions/initial_condition_function.cpp:
```diff
@@ -17,17 +17,253 @@ ::InitialConditionFunction ()
+    // return perturbed vertical velocity component
+    // Reference: Eq.(2.30) -- P. Andersson, L. Brandt, A. Bottaro and D. S. Henningson, "On the breakdown of boundary layer streaks"
+    // Reference for z_scale term: G. LODATO, P. CASTONGUAY AND A. JAMESON, "Discrete filter operators for large-eddy simulation using high-order spectral difference methods", Int. J. Numer. Meth. Fluids (2012)
+    /*const */real F = C*exp(-pow((x-x_loc)/x_scale,2.0))*exp(-pow((y-y_loc)/y_scale,2.0))*cos(beta*(z+half_domain_length_z)/z_scale); // we do (z+half_domain_length_z) because reference has z\in[0,domain_length_z], whereas we center about the z-axis
```
fixed

jbrillon commented on src/dg/strong_dg.cpp:
```diff
@@ -942,6 +950,148 @@ void DGStrong<dim,nstate,real,MeshType>::assemble_volume_term_strong(
+    // -- Solution at legendre poly
+    std::array<std::vector<real>,nstate> legendre_soln_at_q;
+    std::array<dealii::Tensor<1,dim,std::vector<real>>,nstate> legendre_aux_soln_at_q; // legendre auxiliary sol at flux nodes
+    if(this->do_compute_filtered_solution) {
```
I wish but the filtering is embedded in for loops and is difficult to extract as a function, there are similar things being done like this throughout strong_dg.cpp

jbrillon commented on src/dg/strong_dg.cpp:
```diff
@@ -942,6 +950,148 @@ void DGStrong<dim,nstate,real,MeshType>::assemble_volume_term_strong(
+        std::array<std::vector<real>,nstate> primitive_legendre_soln_at_q;
+        std::array<dealii::Tensor<1,dim,std::vector<real>>,nstate> primitive_legendre_aux_soln_at_q; // legendre auxiliary sol at flux nodes
+
+        // Details: this projects to Legendre basis, truncates, then interpolates back to quad nodes.
```
The filtering is done for the velocity field, which is a primitive variable, so that's why we do all the conversion operations

jbrillon commented on src/dg/strong_dg_les.cpp:
```diff
@@ -0,0 +1,1076 @@
+                        }
+                        for (unsigned int idof=0; idof<fe_face_values_extra.dofs_per_cell; ++idof) {
+                            const unsigned int istate = fe_face_values_extra.get_fe().system_to_component_index(idof).first;
+                            soln_at_q[istate] += this->solution[dofs_indices[idof]] * fe_face_values_extra.shape_value_component(idof, iquad, istate);
```
Its only called for advanced SGS models, we can open an issue for it if you'd like

PranshulThakur replied:
That would be great, thanks.

jbrillon replied:
Done, I've opened issue #318 for this!

jbrillon commented on src/dg/weak_dg.cpp:
```diff
@@ -809,6 +808,7 @@ void DGWeak<dim,nstate,real,MeshType>::assemble_boundary_term(
     for (unsigned int iquad=0; iquad<n_quad_pts; ++iquad) {
         const dealii::Tensor<1,dim,real2> normal_int = phys_unit_normal[iquad];
         physics.boundary_face_values (boundary_id, real_quad_pts[iquad], normal_int, soln_int[iquad], soln_grad_int[iquad], soln_ext[iquad], soln_grad_ext[iquad]);
+        physics.boundary_face_values_viscous_flux (boundary_id, real_quad_pts[iquad], normal_int, soln_int[iquad], soln_grad_int[iquad], soln_int[iquad], soln_grad_int[iquad], soln_ext_viscous_flux[iquad], soln_grad_ext_viscous_flux[iquad]);
```
I think the factories will handle this kind of error

PranshulThakur replied:
True, DG factory should handle it.

jbrillon commented on src/dg/weak_dg.cpp:
```diff
@@ -1631,9 +1631,7 @@ void DGWeak<dim,nstate,real,MeshType>::assemble_face_term(
-
-    // Note: This is ignored when use_periodic_bc is set to true -- this variable has no other function when dim!=1
-    if(this->all_parameters->use_periodic_bc == false) {
+    if(this->all_parameters->check_same_coords_in_weak_dg) {
```
At this location in the code, it would print way too many times

PranshulThakur replied:
In that case, maybe we could run this if statement only when `use_periodic_bc == false`. If not, for a periodic case weak DG would print outputs of face coordinates not matching.

jbrillon commented on src/parameters/all_parameters.h:
```diff
@@ -136,10 +140,6 @@ class AllParameters
 
-    ///Flag to use a Classical ESFR scheme where only the surface is reconstructed
-    //The default ESFR scheme is the Nonlinearly Stable FR where the volume is also reconstructed
-    bool use_classical_FR;
```
It was not longer being called in the code after Alex's big PR

jbrillon commented on src/physics/navier_stokes.cpp:
```diff
@@ -55,67 +57,177 @@ NavierStokes<dim, nstate, real>::NavierStokes(
+    // compute tangent vector
+    dealii::Tensor<1,dim,real2> tangent_vector;
+    for(int d=0;d<dim;++d){
+        tangent_vector[d] = velocities_parallel_to_wall[d]/magnitude;
```
If the flow is initialized with non-zero velocity then yes it should never truly be zero

jbrillon commented on src/physics/navier_stokes.cpp:
```diff
@@ -522,6 +799,67 @@ ::compute_viscous_stress_tensor (
+
+template <int dim, int nstate, typename real>
+dealii::Tensor<2,dim,real> NavierStokes<dim,nstate,real>
+::compute_germano_idendity_matrix_M_component (
```
Haha yes it should be! Fixed

cpethrick commented on src/physics/navier_stokes.cpp:
```diff
@@ -37,7 +39,7 @@ NavierStokes<dim, nstate, real>::NavierStokes(
                              two_point_num_flux_type,
                              true,  //has_nonzero_diffusion = true
-                             false) //has_nonzero_physical_source = false
+                             has_nonzero_physical_source) //has_nonzero_physical_source = false
```
I would argue that the comment ` //has_nonzero_physical_source = false` was included when `false` was passed (to make it clear to the user that the parameter has_nonzero_physical_source is the one which is false), therefore it's no longer needed... But it's really a minor thing so feel free to click resolve if you want to leave it as-is

jbrillon replied:
I must've added this for my own clarity when working on this file to not get confused so I think its more help than harm and will leave it as is

jbrillon commented on src/flow_solver/flow_solver_cases/periodic_turbulence.cpp:
```diff
@@ -115,6 +125,14 @@ double PeriodicTurbulence<dim,nstate>::get_constant_time_step(std::shared_ptr<DG
+template <int dim, int nstate>
+unsigned int PeriodicTurbulence<dim,nstate>::get_number_of_degrees_of_freedom_per_state_from_poly_degree(const unsigned int poly_degree_input) const
+{
+    // expression for a uniform grid (i.e. same number of cells in all directions)
```
Whenever the TGV is ran, there is a print statement that tells you details about the grid you're running. For this class it will print `straight_periodic_cube` grid type which is a uniform grid.

jbrillon commented on src/numerical_flux/viscous_numerical_flux.cpp:
```diff
@@ -158,60 +155,61 @@ template<int dim, int nstate, typename real>
+        neighbor_cell_index = current_cell_index;
+        // soln_grad_ext = soln_grad_int;
+        soln_grad_ext = soln_grad_ext_;
+        // filtered_soln_grad_ext = filtered_soln_grad_int;
```
You are right, I guess it was not impacting anything but I've changed line 188 to `filtered_soln_grad_ext = filtered_soln_grad_ext_;` Good catch!

jbrillon commented on src/parameters/all_parameters.cpp:
```diff
@@ -350,6 +362,15 @@ void AllParameters::declare_parameters (dealii::ParameterHandler &prm)
                       "Tolerance for checking that the determinant of surface jacobians at element faces matches. "
                       "Note: Currently only used in weak dg.");
 
+    prm.declare_entry("wall_model_input_from_second_element", "false",// TO DO: change to true for final PR
```
fixed

jbrillon commented on src/physics/model_factory.cpp:
```diff
@@ -122,6 +243,37 @@ ::create_Model(const Parameters::AllParameters *const parameters_input)
+                    parameters_input->two_point_num_flux_type);
+            } else {
+                // Navier-Stokes model does not exist for nstate!=(dim+2) || dim!=3
+                manufactured_solution_function = nullptr;
```
fixed

jbrillon commented on src/testing/turbulent_channel_flow_skin_friction_check.cpp:
```diff
@@ -0,0 +1,355 @@
+    const double percent_emperical_estimate_error = 100.0*abs(computed_skin_friction_coefficient - emperical_estimate_for_skin_friction_coefficient)/emperical_estimate_for_skin_friction_coefficient;
+    pcout << "percent error with computed is " << percent_emperical_estimate_error << " %" << std::endl;
+    if(percent_emperical_estimate_error > 30.0) {
+        pcout << "Warning: considerable difference with emperical estimate for skin friction coefficient value." << std::endl;
```
Nope because we're comparing with an empirical estimate based on the Reynolds number, whereas I check the precise value later on and there I return 1 if the difference is large

jbrillon commented on src/physics/navier_stokes.cpp:
```diff
@@ -893,42 +1366,30 @@ ::boundary_wall (
+    soln_bc[0] = soln_int[0];
+    soln_bc[nstate-1] = soln_int[nstate-1];
+    for (int d=0; d<dim; ++d) {
+        soln_bc[1+d] = -soln_int[1+d];
```
This is documented in my thesis, see below (111 is Brian's thesis): 
<img width="1043" height="563" alt="image" src="https://github.com/user-attachments/assets/188b3cf0-e755-4a79-8d61-2f32845f4d25" />

jbrillon replied:
Added a comment `// Reference: Page 48 of Julien Brillon's thesis available at https://escholarship.mcgill.ca/concern/theses/h989r903p`

jbrillon commented on src/physics/navier_stokes.cpp:
```diff
@@ -893,42 +1366,30 @@ ::boundary_wall (
+        const std::array<real,nstate> primitive_soln_int = this->template convert_conservative_to_primitive_templated<real>(soln_int); // from Euler
+        std::array<real,nstate> primitive_soln_ext = this->template convert_conservative_to_primitive_templated<real>(soln_bc); // from Euler
+        const real temperature_int = this->template compute_temperature<real>(primitive_soln_int);
+        const real temperature_ext = 2.0*this->isothermal_wall_temperature - temperature_int;
```
For temperature, it becomes linear with the no slip BC since v=0 and since we solve for total energy, `E=rho*Cv*T+rho*v*v/2` we are left with `E=rho*Cv*T`

jbrillon commented:
My sincere apologies to everyone for the rather long wait on getting all of this completed, however all comments have now been addressed! I am completing issue #317 right now, and then will merge the latest master, and re-run ctest and post the results here before merging!
